### PR TITLE
TST: Use tox for CI

### LINKED ISSUE #1054: TST: Test against numpy 2.0.dev
numpy 2.0.dev is available from their nightly wheel and it has breaking changes. Since the backend here calls a lot of numpy, better make sure they are compatible sooner than later.

### PR DESCRIPTION
**Pull request by pllim**
Use tox for CI except for the conda job.

Fix #1054

### CODE REVIEW

**Comment by pllim** [comment]
@ejeschke , this PR switches the test machinery to tox. This makes it easier to mix and match test matrix and also pulling in different upstream dev wheels (e.g., numpy-dev).

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,93 @@
+[tox]
+envlist =
+    py{38,39,310,311}-test{,-oldestdeps,-devdeps,-predeps}
+    codestyle
+    twine
+    bandit
+requires =
+    setuptools >= 30.3.0
```
Comment (ejeschke) [inline tox.ini:8]: No need for `setuptools_scm` in `tox.ini`?

**Review by ejeschke — APPROVED** [review]
@pllim, I'm new to `tox` (but learning!) so will trust to your judgement on this one.  It sounds like it will give some added flexibility. FFTM

File: tox.ini (hunk 1 above)
Comment (pllim) [reply tox.ini:8]: `astropy` does not list it, and it is already declared in pyproject.toml . But if you want it listed, doesn't hurt. 🤷

Comment (ejeschke) [reply tox.ini:8]: Only if an accurate version number is needed in the testing?

Comment (pllim) [reply tox.ini:8]: Maybe. I actually don't fully understand this section. I was just copy pasting from a setup I know will work. Do you want me to remove it and see?

Comment (ejeschke) [reply tox.ini:8]: I guess if tox does not bypass the `setup.cfg` then it shouldn't be necessary ...?

Comment (ejeschke) [reply tox.ini:8]: > Do you want me to remove it and see?

I think we need it in `setup.cfg` (and likely `pyproject.toml`). I just don't know how `tox` interacts with those, but presumably uses pip to install things so I suppose that is good enough.  So no need to change anything here, I guess.

Comment (pllim) [reply tox.ini]: I removed it. I am curious now.

**Comment by pllim** [comment]
CI still passed. I guess we don't need that section after all, good catch!

**Comment by ejeschke** [comment]
Thanks for investigating. FFTM

**Comment by pllim** [comment]
Thanks for the review!

**Pull request merged**